File: Single-diode model/tracer_single_diode_emulator.m
clc; clear; close all; format long

%% 1) Module Parameters

ns = 60; % Number of series cells
np = 1; % Number of parallel branches

VMPmod = 30.1; % Voltage at maximum power point of the module (V)
IMPmod = 8.30; % Current at maximum power point of the module (A)
VOCmod = 37.2; % Open-circuit voltage of the module (V)
ISCmod = 8.87; % Short-circuit current of the module (A)

Vmp_cell = VMPmod / ns;
Imp_cell = IMPmod / np;
Voc_cell = VOCmod / ns;
Isc_cell = ISCmod / np;

Tref = 25 + 273.15; % Reference temperature (K)
Sref = 1000; % Reference irradiance (W/m²)

muICC = 0.0058; % Temperature coefficient
%% 2) Operating Conditions

T = 44.5 + 273.15; % Current temperature (K)
S = 765; % Current irradiance (W/m²)

%% 3) Physical Constants

q = 1.60217662e-19; % Elementary charge (C)
k = 1.38064852e-23; % Boltzmann constant (J/K)
Eg = 1.21 * q; % Silicon band gap energy (J)

%% 4) Parameter Estimation via fsolve

x0 = [Isc_cell; 1e-9; 3; 0.01; 2];
opts = optimoptions('fsolve', ...
    'Display', 'off', 'TolFun', 1e-10, 'TolX', 1e-10, 'MaxIter', 1000, 'MaxFunctionEvaluations', 2000);
fun = @(x) residuals_2_20(x, Voc_cell, Isc_cell, Vmp_cell, Imp_cell, q, k, Tref);
[xsol, ~, exitflag] = fsolve(fun, x0, opts);
if exitflag <= 0
    warning('fsolve did not converge (exitflag = %d)', exitflag);
end

Iph_ref = xsol(1);
Is0_ref = xsol(2);
A = xsol(3);
Rs = xsol(4);
Rp = xsol(5);

fprintf('Iph_ref = %.6f A\n', Iph_ref);
fprintf('Is0_ref = %.2e A\n', Is0_ref);
fprintf('A       = %.6f\n', A);
fprintf('Rs      = %.6f Ohms\n', Rs);
fprintf('Rp      = %.6f Ohms\n', Rp);

%% Characteristic Points (used to build the I-V model)

V_cell = [
    0.0;
    0.2 * VMPmod / ns;
    0.4 * VMPmod / ns;
    0.6 * VMPmod / ns;
    0.8 * VMPmod / ns;
    0.9 * VMPmod / ns;
    VMPmod / ns;
    (VMPmod / ns + VOCmod / ns) / 2.0;
    0.9 * VOCmod / ns;
    0.95 * VOCmod / ns;
    VOCmod / ns];

I_cell = arrayfun(@(V) solve_I_V_2_11(V, Iph_ref, Is0_ref, A, Rs, Rp, ...
    q, k, Eg, S, Sref, muICC, T, Tref), V_cell);
points_V = V_cell * ns;
points_I = I_cell * np;

%% Piecewise PV Model

% Coefficient initialization
coeffs = zeros(length(points_V) - 1, 2);
for i = 1:length(points_V) - 1
    a = (points_I(i + 1) - points_I(i)) / (points_V(i + 1) - points_V(i));
    b = points_I(i) - a * points_V(i);
    coeffs(i, :) = [a, b]; % Store [a, b] in each row
end

% MATLAB function for the PV model (accepts a voltage vector)
modele_pv = @(V) piecewise_pv_model(V, points_V, coeffs);
function I_out = piecewise_pv_model(V_in, points_V, coeffs)
    I_out = zeros(size(V_in)); % Initialize the output current array
    for k = 1:numel(V_in) % Iterate over each voltage input
        V = V_in(k);
        found = false;
        for i = 1:length(points_V) - 1
            % Use a small tolerance for floating-point comparisons
            tol = 1e-9;
            if (points_V(i) - tol <= V && V <= points_V(i + 1) + tol)
                a = coeffs(i, 1);
                b = coeffs(i, 2);
                I_out(k) = a * V + b;
                found = true;
                break;
            end
        end
        if ~found
            I_out(k) = 0; % Default to 0 if V is outside defined segments
        end
    end
end

%% Provided Data: (R, V_test, I_test, V*, I*)

mesures = [
    % ---------------------KC200GT---------------------
    
    % --- Sref = 1000, Tref = 25, S = 511, T = 54,3 ---
    % 3.409547739,30.042597,9.221842,13.882890,4.261477;
    % 3.837837838,29.833204,7.930077,15.970324,4.245132;
    % 4.282878412,30.036655,7.134727,17.581207,4.176135;
    % 4.674074074,29.913689,6.374939,19.258860,4.104277;
    % 5.430446194,29.912422,5.737882,21.006065,4.029441;
    % 6.206703911,29.915359,4.927178,22.505482,3.706742;
    % 7.051359517,30.027420,4.394181,23.636152,3.458890;
    % 8.111864407,29.946547,3.833536,24.225838,3.101213;
    % 11.20089286,29.959095,2.824838,25.353563,2.390584;
    % 15.69277108,29.962294,2.051082,26.291145,1.799772;
    % 34.67532468,29.970604,1.060586,26.867710,0.950782;
    % 405.9701493,29.976196,0.295321,27.326483,0.269217;

    % --- Sref = 511, Tref = 54,3, S = 511, T = 54,3 ---
    % 3.298200514,30.083759,9.238234,13.129311,4.031798;
    % 4.58778626,29.899143,6.469102,18.305553,3.960665;
    % 5.436997319,29.930714,5.628777,20.560305,3.866576;
    % 5.991689751,29.945515,5.095525,21.877769,3.722718;
    % 6.424068768,29.967121,4.802621,22.681322,3.634977;
    % 7.182098765,29.964804,4.303484,23.515127,3.377195;
    % 8.49825784,29.994026,3.657519,24.605215,3.000399;
    % 9.86328125,29.990040,3.184377,25.463463,2.703740;
    % 12.61927945,30.060379,2.518887,26.104248,2.187386;
    % 16.58125,30.009392,1.938024,26.694448,1.723943;
    % 31.30484988,30.022461, 1.095335,27.238712,0.993773;
    % 65.39379475,30.024586,0.648094,27.517773,0.593983;
    % 40.69117647,30.027824,0.280485,27.751440,0.259222;

    % ---------------------KC85TS---------------------
    
    % --- Sref = 1000, Tref = 25, S = 1000, T = 25 ---
    % 2.560636183,23.837536,9.623577,13.199326,5.328769;
    % 2.747572816,23.819389,8.729925,14.48859,5.31014;
    % 2.943359375,23.843975,8.196189,15.374142,5.284747;
    % 3.305498982,23.867809,7.438157,16.523491,5.149376;
    % 3.544885177,23.967472,6.982988,17.287716,5.036823;
    % 4.417475728,23.891666,5.552080,18.478067,4.294037;
    % 5.591304348,23.895893,4.412038,19.506063,3.601518;
    % 7.574144487,23.901066,3.292860,20.110668,2.770655;
    % 12.26190476,23.910973,2.129044,20.725580,1.845415;
    % 31.58208955,23.911114,0.999190,21.232008,0.887237;
    % 405.8490566,23.910839,0.326262,21.545561,0.293988;

    % -------------- KC85TS - charge R+L -------------- 
    
    % --- Sref = 1000, Tref = 25, S = 1000, T = 25 ---
    % 2.291750503,23.776310,10.092688,12.558353,5.330833;
    % 2.628458498,23.813004,8.758986,14.440410,5.311524;
    % 2.784046693,23.824802,8.113836,15.506546,5.280949;
    % 2.940828402,23.840160,7.710728,16.109459,5.210353;
    % 3.298780488,23.937588,6.910994,17.392550,5.021383;
    % 3.82735426,23.850681,5.991889,18.101677,4.547595;
    % 4.816489362,23.860895,4.968657,18.984102,3.953141;
    % 5.238095238,23.863398,4.379676,19.531202,3.584584;
    % 7.25,23.871922,3.343837,20.080969,2.812823;
    % 10.32307692,23.874191,2.385726,20.611444,2.059683;
    % 21.25510204,23.875790, 1.307175,21.090651, 1.154691;
    % 406.0377358,23.883148,0.306018,21.554970,0.276186;

    % ---------------------CS6P-250P---------------------
    
    % --- Sref = 1000, Tref = 25, S = 556, T = 33 ---
    % 4.966173362,36.916618,7.636598,23.861177,4.935940;
    % 5.235789474,36.925896,7.135579,25.287460,4.886562;
    % 5.670940171,36.925285,6.611194,26.927507,4.821167;
    % 6.189309577,36.964649,6.070982,28.176317,4.627609;
    % 7.035799523,37.076538,5.380336,29.874315,4.335190;
    % 8.924637681,36.992016,4.264987,31.124376,3.588478;
    % 12.23137255,37.003716,3.019732,32.481518,2.650693;
    % 18.71348315,37.020912,2.094456,33.569107,1.899170;
    % 49.46376812,37.018486,0.900303,34.318546,0.834639;
    % 406.4705882,37.029892,0.242107,34.741459,0.227145;

    % --- Sref = 1000, Tref = 25, S = 765, T = 44.5 ---
    3.276853253,36.058861,11.093060,22.148668,6.813763;
    3.473604827,35.893806,10.355574,23.533447,6.789537;
    3.797527048,35.901566,9.535309,25.083090,6.661966;
    4.232854864,36.052902,8.577839,27.035177,6.432309;
    4.633445946,35.951599, 7.810769,27.919050,6.065634;
    5.251838235,35.942787,6.917910,29.022854,5.586030;
    6.098138748,36.042290,5.907085,30.229673,4.954437;
    13.33891213,35.980782, 2.834330,32.149712,2.532543;
    61.90740741,35.987362,0.760594,33.594578,0.710022;
    421.125,35.992996,0.285993,33.809090,0.268640;

    % ---------------Uni-Solar ES-62T-----------------
    
    % --- Sref = 1000, Tref = 25, S = 1000, T = 25 ---
    % 2.68852459,22.999146,8.928772,11.900667,4.620100;
    % 2.795348837,23.152687,8.527935,12.389590,4.563514;
    % 3.073286052,23.148125,7.678766,13.375956,4.437113;
    % 3.569620253,23.013630,6.702841,14.468855,4.214130;
    % 3.780927835,23.023594,6.254211,15.035897,4.084404;
    % 5.068535826,23.038422,4.691563,16.597288,3.379885;
    % 7.243902439,23.059406,3.418242,18.088655,2.681396;
    % 9.658031088,23.066004,2.557652,18.883083,2.093833;
    % 20.5625,23.081512,1.335646,19.927521,1.153136;
    % 92.95454545,23.119362,0.512176,20.578976,0.455897;
    ];

if ~isempty(mesures)
    R_data = mesures(:, 1);
    V_violet = mesures(:, 2); % Purple Points (Test Point)
    I_violet = mesures(:, 3); % Purple Points (Test Point)
    V_noir = mesures(:, 4);   % Black Points (Intersection)
    I_noir = mesures(:, 5);   % Black Points (Intersection)
else
    R_data = []; V_violet = []; I_violet = []; V_noir = []; I_noir = [];
end
%% Plotting (Figure Generation)

figure; 
hold on; 

% 1. Plot "Load Lines" FIRST
if isempty(R_data)
    Rs_to_plot = [0.1, 0.5, 1:1:15, 20:5:50, 60:10:200, 300, 400, 500, 1000];
    Rs_to_plot = sort([Rs_to_plot, inf]);
else
    Rs_to_plot = R_data; % Use resistances from your data
    Rs_to_plot = sort(unique(Rs_to_plot));
end
V_line = linspace(0, max(V_violet) * 1.25, 100);
for r_val = Rs_to_plot'
    if isinf(r_val) % Open circuit (I = 0)
        plot(V_line, zeros(size(V_line)), 'k--', 'LineWidth', 0.8, 'Color', [0.5 0.5 0.5], 'HandleVisibility', 'off');
    elseif r_val == 0 % Short circuit (V = 0)
        plot(zeros(size(V_line)), linspace(0, max(I_violet) * 1.1, 100), 'k--', 'LineWidth', 0.8, 'Color', [0.5 0.5 0.5], 'HandleVisibility', 'off');
    else
        I_line = V_line ./ r_val;
        plot(V_line, I_line, 'k--', 'LineWidth', 0.8, 'Color', [0.5 0.5 0.5], 'HandleVisibility', 'off');
    end
end

% 2. Plot the I-V Model (P-V Curve)
V_plot = linspace(0, VOCmod, 500);
I_plot = modele_pv(V_plot);
h_model = plot(V_plot, I_plot, 'r-', 'LineWidth', 2, 'DisplayName', 'I-V Curve - Single diode model');

% 3. Plot "Test Points" (Purple Points)
h_test_point = []; 
if ~isempty(V_violet)
    h_test_point = scatter(V_violet, I_violet, 40, 'm', 'filled', 'DisplayName', 'Test Point');
end

% 4. Plot "Intersections" (Black Points)
h_intersection = []; 
if ~isempty(V_noir) && ~isempty(I_noir)
    h_intersection = scatter(V_noir, I_noir, 70, 'ks', 'filled', 'DisplayName', 'Intersection');
end

%% Plot Configurations

xlabel('Voltage (V)', 'FontSize', 30); 
ylabel('Current (A)', 'FontSize', 30);

% Axis limits to match the provided figure
xlim([0 1.2 * VOCmod]);
ylim([0 1.2 * max(I_violet)]);

% Font configurations for axis ticks
set(gca, 'FontSize', 30); 
grid off;

% Create the legend and adjust position and font
legend_handles = [h_model];
legend_labels = {'I-V Curve - Single diode model'};

if ~isempty(h_test_point) && ishandle(h_test_point)
    legend_handles = [legend_handles, h_test_point];
    legend_labels = [legend_labels, 'Test Point'];
end
if ~isempty(h_intersection) && ishandle(h_intersection)
    legend_handles = [legend_handles, h_intersection];
    legend_labels = [legend_labels, 'Intersection'];
end

leg = legend(legend_handles, legend_labels, 'Location', 'NorthWest', 'FontSize', 20);

set(leg, 'Box', 'on'); 
set(findall(gcf, '-property', 'FontName'), 'FontName', 'Times New Roman');
set(findall(gca, '-property', 'FontName'), 'FontName', 'Times New Roman');
hold off;

%% --- Auxiliary Functions ---

function F = residuals_2_20(x, Voc, Isc, Vmp, Imp, q, k, Tref)
    Iph = x(1); Is0 = x(2); A = x(3); Rs = x(4); Rp = x(5);
    C = q / (A * k * Tref);
    cap = 700;
    a_sc = min(C * Rs * Isc, cap);
    a_oc = min(C * Voc, cap);
    a_mp = min(C * (Rs * Imp + Vmp), cap);
    F = zeros(5, 1);
    F(1) = Isc - (Iph - Is0 * (exp(a_sc) - 1) - (Rs * Isc) / Rp);
    F(2) = 0 - (Iph - Is0 * (exp(a_oc) - 1) - (Voc) / Rp);
    F(3) = Imp - (Iph - Is0 * (exp(a_mp) - 1) - (Rs * Imp + Vmp) / Rp);
    F(4) = Rs + (q * Is0 * Rp * (Rs - Rp) / (A * k * Tref)) * exp(a_sc);
    term = 1 + q * (Vmp - Rs * Imp) / (A * k * Tref);
    F(5) = Iph - 2 * Vmp / Rp - (Is0 * term * (exp(a_mp) - 1));
end

function I = solve_I_V_2_11(V, Iph_ref, Is0_ref, A, Rs, Rp, ...
    q, k, Eg, S, Sref, muICC, T, Tref)
    Iph = Iph_ref * (S / Sref) + muICC * (T - Tref);
    gamma = (T / Tref)^3 * exp(Eg / (A * k) * (1 / Tref - 1 / T));
    I = Is0_ref * gamma; % Initial estimate
    cap = 700;
    for it = 1:30
        C = q / (A * k * T);
        a = min(C * (Rs * I + V), cap);
        expo = exp(a);
        f = Iph ...
            - (Rs * I + V) / Rp ...
            - Is0_ref * gamma * (expo - 1) ...
            - I;
        df = -Rs / Rp ...
            - Is0_ref * gamma * expo * (q * Rs / (A * k * T)) ...
            - 1;
        dI = -f / df;
        I = I + dI;
        if abs(dI) < 1e-6, break; end
    end
end

File: Single-diode model/Plotting Algorithms/Python/tracer_single_diode_emulator.py
import numpy as np
from scipy.optimize import least_squares
import matplotlib.pyplot as plt
import warnings

# --- Auxiliary Functions ---

def residuals_2_20(x, Voc, Isc, Vmp, Imp, q, k, Tref):
    """
    Calculates the residuals for the five-point model equations.
    This function is used by least_squares to find the model parameters.
    """
    Iph, Is0, A, Rs, Rp = x

    C = q / (A * k * Tref)
    cap = 700  # Safety cap to prevent overflow in exp()

    # Arguments for the exponential function, capped for stability
    a_sc = min(C * Rs * Isc, cap)
    a_oc = min(C * Voc, cap)
    a_mp = min(C * (Rs * Imp + Vmp), cap)

    F = np.zeros(5)
    F[0] = Isc - (Iph - Is0 * (np.exp(a_sc) - 1) - (Rs * Isc) / Rp)
    F[1] = 0 - (Iph - Is0 * (np.exp(a_oc) - 1) - Voc / Rp)
    F[2] = Imp - (Iph - Is0 * (np.exp(a_mp) - 1) - (Rs * Imp + Vmp) / Rp)
    F[3] = Rs + (q * Is0 * Rp * (Rs - Rp) / (A * k * Tref)) * np.exp(a_sc)
    term = 1 + q * (Vmp - Rs * Imp) / (A * k * Tref)
    F[4] = Iph - 2 * Vmp / Rp - (Is0 * term * (np.exp(a_mp) - 1))
    
    return F

def solve_I_V_2_11(V, Iph_ref, Is0_ref, A, Rs, Rp,
                   q, k, S, Sref, alpha, T, Tref, E_G0, k1, k2):
    """
    Solves for the cell current (I) at a given voltage (V) using Newton-Raphson.
    """
    # 1. Photocurrent with T and S dependency
    Iph = (S / Sref) * (Iph_ref + alpha * (T - Tref))
    
    # 2. Temperature-dependent Band Gap Energy (Varshni's equation)
    Eg_ref_J = (E_G0 - (k1 * Tref**2) / (Tref + k2)) * q
    Eg_J = (E_G0 - (k1 * T**2) / (T + k2)) * q
    
    # 3. Saturation Current with T and Eg(T) dependency
    exponent_term = (Eg_ref_J / (A * k * Tref)) - (Eg_J / (A * k * T))
    Is = Is0_ref * (T / Tref)**3 * np.exp(exponent_term)
    
    # 4. Iterative solver (Newton-Raphson)
    I = Iph  # Initial guess
    for _ in range(30): # Iterate up to 30 times
        V_diode = V + I * Rs
        arg_exp = min(q * V_diode / (A * k * T), 700) # Capped for stability
        expo = np.exp(arg_exp)
        
        f = Iph - Is * (expo - 1) - V_diode / Rp - I
        df = -Is * expo * (q * Rs / (A * k * T)) - Rs / Rp - 1
        
        # Avoid division by zero
        if abs(df) < 1e-9:
            break
            
        dI = -f / df
        I = I + dI
        
        if abs(dI) < 1e-6:
            break
            
    return I

def piecewise_pv_model(V_in, points_V, points_I):
    """
    Calculates the current from a piecewise linear model using interpolation.
    """
    # np.interp performs piecewise linear interpolation.
    # It requires the x-coordinates (points_V) and y-coordinates (points_I).
    # `left` and `right` define values for inputs outside the interpolation range.
    I_out = np.interp(V_in, points_V, points_I, left=0.0, right=0.0)
    return I_out

# --- Main Script ---

## Module Parameters
ns = 60       # Number of series cells
n_p = 1       # Number of parallel branches 

Vmp_mod_ref = 30.1 # Voltage at maximum power point of the module (V)
Imp_mod_ref = 8.30 # Current at maximum power point of the module (A)
Voc_mod_ref = 37.2 # Open-circuit voltage of the module (V)
Isc_mod_ref = 8.87 # Short-circuit current of the module (A)

Tref = 25 + 273.15  # Reference temperature (K)
Sref = 1000         # Reference irradiance (W/m²)
alpha = 0.00065     # temperature coefficient of Isc (%/K)

## Operating Conditions
T = 44.5 + 273.15   # Current temperature (K)
S = 765           # Current irradiance (W/m²)

## PV cell parameters 
Vmp_cell_ref = Vmp_mod_ref / ns
Imp_cell_ref = Imp_mod_ref / n_p 
Voc_cell_ref = Voc_mod_ref / ns
Isc_cell_ref = Isc_mod_ref / n_p 

## Physical Constants
q = 1.60217662e-19  # Elementary charge (C)
k = 1.38064852e-23  # Boltzmann constant (J/K)
E_G0 = 1.166          # Band gap energy at 0K (eV)
k1 = 4.73e-4          # Coefficient k1 (eV/K)
k2 = 636              # Coefficient k2 (K)

## 4) Parameter Estimation via least_squares (more robust than fsolve)
# Initial guess based on MATLAB's x0 for better convergence
x0 = np.array([Isc_cell_ref, 1e-9, 3.0, 0.01, 2.0])

# Define physical bounds for the parameters ([Iph, Is0, A, Rs, Rp])
lb = [0, 0, 0.5, 0, 1e-3]  # Lower bounds
ub = [Isc_cell_ref * 1.2, 1e-6, 5.0, 1.0, 10000] # Upper bounds

result = least_squares(
    residuals_2_20, 
    x0, 
    args=(Voc_cell_ref, Isc_cell_ref, Vmp_cell_ref, Imp_cell_ref, q, k, Tref), 
    bounds=(lb, ub),
    method='trf', # Trust Region Reflective algorithm, good for bounds
    ftol=1e-10,
    xtol=1e-10,
    gtol=1e-10
)

if not result.success:
    warnings.warn(f'least_squares did not converge. Status: {result.status}, Message: {result.message}')

Iph_ref, Is0_ref, A, Rs, Rp = result.x
print(f'Iph_ref = {Iph_ref:.6f} A')
print(f'Is0_ref = {Is0_ref:.2e} A')
print(f'A       = {A:.6f}')
print(f'Rs      = {Rs:.6f} Ohms')
print(f'Rp      = {Rp:.6f} Ohms')

## 5) Characteristic Points (used to build the I-V model)
V_cell = np.array([
    0.0,
    0.2 * Vmp_mod_ref / ns,
    0.4 * Vmp_mod_ref / ns,
    0.6 * Vmp_mod_ref / ns,
    0.8 * Vmp_mod_ref / ns,
    0.9 * Vmp_mod_ref / ns,
    Vmp_mod_ref / ns,
    (Vmp_mod_ref / ns + Voc_mod_ref / ns) / 2.0,
    0.9 * Voc_mod_ref / ns,
    0.95 * Voc_mod_ref / ns,
    Voc_mod_ref / ns
])

# Use a list comprehension, which is a Pythonic way to apply a function to each element
I_cell = np.array([solve_I_V_2_11(v, Iph_ref, Is0_ref, A, Rs, Rp,
                                q, k, S, Sref, alpha, T, Tref, E_G0, k1, k2) for v in V_cell])

points_V = V_cell * ns
points_I = I_cell * n_p # Using the corrected variable n_p

## 6) Piecewise PV Model (Coefficient calculation is no longer needed)
# The model function now uses points_V and points_I directly
modele_pv = lambda V: piecewise_pv_model(V, points_V, points_I)


## 7) Provided Data: (R, V_test, I_test, V*, I*)
mesures = np.array([
    
    # Data from CS6P-250P under S = 556, T = 33
    # [3.312101911, 35.098381, 10.830381, 15.96633, 4.926764],
    # [4.470212766, 35.093315, 8.052373, 21.384974, 4.906911],
    # [5.072649573, 34.928596, 7.081912, 24.119793, 4.890384],
    # [5.428571429, 34.964813, 6.637497, 25.479982, 4.836957],
    # [5.846827133, 34.940865, 6.137167, 27.134899, 4.766093],
    # [6.560465116, 34.992397, 5.522508, 28.611193, 4.515424],
    # [7.335802469, 35.004902, 4.953645, 30.103748, 4.260068],
    # [11.3297491, 35.064438, 3.307633, 31.929382, 3.011903],
    # [20.85, 35.042049, 1.906419, 33.620697, 1.829092],
    # [62.91512915, 35.042969, 0.812353, 34.34491, 0.796171],
    
    
    # Data from CS6P-250P under S = 765, T = 44.5 
     [3.424196018,34.855976,10.283596,22.807545,6.728934],
     [3.684782609,34.869011,9.658541,24.17536,6.696453],
     [4.031847134,34.885811,8.816074,25.775291,6.513734],
     [4.44278607,34.931274,8.078631,27.240286,6.29992],
     [5.227356747,34.942135,6.905494,28.715485,5.674942],
     [5.995125914,34.961349,6.021228,29.939571,5.156351],
     [7.701643489,34.964321,4.749564,30.825687,4.187371],
     [13.45162653,34.977104,2.806052,32.122658,2.577053],
     [68.56557377,35.020119,0.753771,33.615498,0.723538],
])

# Use Python's 0-based indexing
R_data = mesures[:, 0]
V_violet = mesures[:, 1]  # Purple Points (Test Point)
I_violet = mesures[:, 2]  # Purple Points (Test Point)
V_noir = mesures[:, 3]    # Black Points (Intersection)
I_noir = mesures[:, 4]    # Black Points (Intersection)

## 8) Plotting (Figure Generation)
plt.style.use('default') # Use a standard style
fig, ax = plt.subplots(figsize=(12, 8))

# 1. Plot "Load Lines"
Rs_to_plot = np.sort(np.unique(R_data))
V_line_max = np.max(V_violet) * 1.25 if len(V_violet) > 0 else Voc_mod_ref * 1.1
V_line = np.linspace(0, V_line_max, 100)

for r_val in Rs_to_plot:
    if r_val > 0 and np.isfinite(r_val):
        I_line = V_line / r_val
        # Corrected plot call to avoid warning
        ax.plot(V_line, I_line, linestyle='--', linewidth=0.8, color='0.5')

# 2. Plot the I-V Model
V_plot = np.linspace(0, Voc_mod_ref, 500)
I_plot = modele_pv(V_plot)
h_model, = ax.plot(V_plot, I_plot, 'r-', linewidth=2, label='I-V Curve - Single diode model')

# 3. Plot "Test Points" (Purple Points)
h_test_point = None
if V_violet.size > 0:
    h_test_point = ax.scatter(V_violet, I_violet, s=40, c='m', marker='o', label='Test Point', zorder=5)

# 4. Plot "Intersections" (Black Points)
h_intersection = None
if V_noir.size > 0 and I_noir.size > 0:
    h_intersection = ax.scatter(V_noir, I_noir, s=70, c='k', marker='s', label='Intersection', zorder=5)

## 9) Plot Configurations
plt.rcParams['font.family'] = 'serif'
plt.rcParams['font.serif'] = 'Times New Roman'
plt.rcParams['axes.labelweight'] = 'normal'

ax.set_xlabel('Voltage (V)', fontsize=22)
ax.set_ylabel('Current (A)', fontsize=22)
ax.set_xlim(0, 1.01 * Voc_mod_ref)
# Use max of measured current or calculated short-circuit current for robust y-limit
ylim_max = max(np.max(I_violet) if I_violet.size > 0 else 0, Isc_mod_ref)
ax.set_ylim(0, 1.1 * ylim_max)
ax.tick_params(axis='both', which='major', labelsize=20)
ax.grid(False)

# Create the legend
handles = [h_model]
if h_test_point:
    handles.append(h_test_point)
if h_intersection:
    handles.append(h_intersection)

leg = ax.legend(handles=handles, loc='upper right', fontsize=16) # 'NorthWest' in MATLAB is often 'upper left' or 'upper right' in Matplotlib
leg.get_frame().set_edgecolor('k') # Add box around legend
leg.get_frame().set_linewidth(1.0)

plt.tight_layout()
plt.show()

